cbv3_django_prototype/cbv3_django_prototype/resources/models.py

```diff
@@ -1,3 +1,47 @@
+import uuid
+import django.contrib.postgres.fields as pg
 from django.db import models
+from django.utils import timezone
 
-# Create your models here.
+class Resource(models.Model):
+
+    id = models.UUIDField(primary_key=True, default=uuid.uuid1, editable=False)
+
+    title = models.CharField(max_length=200)
+
+    #potentially a markdown field.  Will need a markdown converter and renderer
+    description = models.TextField(blank=True, max_length=500)
+
+    #specific URL of resource
+    url = models.URLField(max_length=300)
+
+    #the URL of the referring/source site.  e.g. URL of tweet, if it was tweeted.
+    referrer = models.URLField(blank=True, max_length=300)
+
+    #names of persons who suggest this resource to the user
+    credit = models.CharField(max_length=100)
+
+    # publication date of resource
+    date_published = models.DateTimeField(default=timezone.now)
+
+    # creation date of resource entry
+    created = models.DateTimeField(auto_now_add=True)
+
+    # modification date of resource entry
+    modified = models.DateTimeField(default=timezone.now)
+
+    # [ video, podcast, talk, tutorial, course, book, blog ]
+    type = models.CharField(max_length=100)
+
+    paid = models.BooleanField(null=True)
+
+    # JSON here, so we can parse them easier....
+    tags = pg.JSONField()
```

Comment:
I think tags should be a `ManyToMany` relationship, to a `Tags` table. Serialization/deserialization is relatively easy with DRF.

Reply:
The idea is to use relational capabilities of the database first, before resorting to storing nested data. JSON should be used sparingly, if at all, and for specific optimixatiins, like when joins might be too slow.

Using a proper ManyToMany approach here might have other benefits like allowing for autocomplete or clustering related resources (like a recommendation engine).

Reply:
Agreed.  The classic way to model something like this would be with a many-to-many and a Tags table.  This was more a quick nod to the API specs, as well as an experiment with postgres JSONB capabilities.  So a bit of a hand-wave.

But as this system grows, I do have a few concerns/caveats with a "simple" many-to-many Tags table:

1.  Tags will most likely be a freeform field on the front end.  I don't really see our users choosing from an epic list of tags...and the "common convention" is for users to enter what they see fit.  This means writing a bunch of code to search for already existing tags and similar tags, and only add truly new/unique tags to the DB.  Even with that, the Tags table will potentially be a mess of overlapping and duplicate data from the get-go.  Just mocking up 28 resources for testing data created a "mess" -- and I am only one human.

2.  Learning Paths, Hangouts, Notes and Discussions will also be tagged, I would imagine.  So there will be potentially a lot of relations tracing back to the Tags table, each having overlapping, but not necessarily identical topics....so each would have to access the code for searching for existing tags/adding new tags.  With virtually everything connecting to the Tags table (and it being updated a lot per user, hangout, note, discussion, etc.), it could become a bottleneck fairly quickly.

3.  Postgres can index JSONB, and perform many of the same queries on JSONB that it can on regular data with small syntax changes.   Both autocomplete and recommendations can be done (albeit not at top speed) with documents.  But yes - if we are going the RDBMS route, then we shouldn't use JSONB too often, or in too many places.

One option is to use https://github.com/jazzband/django-taggit along with https://github.com/glemmaPaul/django-taggit-serializer, and customize the code so that each model has it's tags.  See this for a rundown and some thoughts:  https://medium.com/sthzg/a-short-exploration-of-django-taggit-bb869ea5051f.

And all that being said, I don't like how the Tags field adds an extra level right now, so I'm going to look at re-writing the serializer to omit that, and do some more digging on how Tags are implemented in other systems to see some more options.  Do you have a go-to model/resource you can share??  That would be great.

Reply:
We ended up using a lot of JSONB where I work, and it has been a bit of a wrestling match. The main take-away, from a senior developer colleague of mine, is to only use JSONB for specific optimizations and to default to using a relational model with Postgres.

As mentioned, we can use an auto-complete widget along with the django-taggit package to improve consistency and avoid bespoke tagging code.

It might be worth checking out how Wagtail handles tagging, as it is a well-thought-out CMS:

https://docs.wagtail.io/en/v2.6.2/reference/pages/model_recipes.html
https://learnwagtail.com/tutorials/adding-tags-to-wagtail-cms-pages/

Reply:
I've re-written the tag field so that it doesn't have the weird extra level, but I am going to leave it as JASONB in the DB for now for several reasons:

1)  It is not clear in my reading & research that we'd take any sort of performance hit with using JASON in this fashion, and in fact several sources show JSONB to be preferable for speed and ease of use in a tagging model.

2)  This is currently a highly simplified model of what might be implemented in the DB should we go the Django/DRF route.  At this time, I don't necessarily want to complicate matters by adding 3 more tables to the mix, and JASONB gets our point across without them.

3)  I don't know that we have clarity around the importance/meaning/use of tags in our system, and I'd like to have some further discussions with others who are doing prototypes to hear how they envision it working.

4)  Changing the model a little later is not going to be terribly painful, since we are still in a prototyping phase.

Reply:
Just so  that it is connected to this pull request, here is the JSON for the tags field:

"tags": [
                {
                    "id": "a1167632-abb8-4ce1-b090-b84e79c94c7f",
                    "name": "JavaScript"
                },
                {
                    "id": "66cc3539-6705-45ee-b15d-644d5fb90c41",
                    "name": "ES6"
                },
                {
                    "id": "8b99bcd9-a9a5-4455-b2da-fed06051d502",
                    "name": "React"
                }
            ]

Reply:
To expand a little bit, JSONB is good to used when we need 'data locality', in effect avoiding a second query or JOIN to get a related entity.

In general, Tags are used to show related content, e.g. "resources tagged Python". I believe that it will be desirable to use these tags to create views showing similar items, resources, etc.

The performance penalty comes when we try to search within the JSONB column (even with an index) for a tag string in the `name` property, and then join related documents together.

For these reasons, JSONB does not seem like a good approach for tagging, particularly when [the django taggit package makes a proper M2M tagging approach very straightforward](https://django-taggit.readthedocs.io/en/latest/getting_started.html).

Please excuse my directness, but I strongly recommend we use the django-taggit approach here or to omit the tagging feature from this pull request.

Reply:
Let's leave tags implementation for a future issue -- I'll file @brylie's comments about using the `taggit` package (looks interesting!) and the performance penalty point makes sense.

> I don't know that we have clarity around the importance/meaning/use of tags in our system, and I'd like to have some further discussions with others who are doing prototypes to hear how they envision it working.

> Changing the model a little later is not going to be terribly painful, since we are still in a prototyping phase.

Agree -- it's not terrible if our model isn't perfect right now, as long as we don't forget to change it if we decide on continiuing with relational/postgres after the architectural comparison.

I think this PR is ready to merge in, but if there are other issues that should be filed (that I missed), please feel free to do so! Thanks again, all.

Reply:
Also sorry @BethanyG, was scrolling back and realized you mentioned the `django-taggit` package first! That's what I get for skimming. :) Anyway, have filed the issue to follow up on tags when we need to here: https://github.com/codebuddies/django-concept/issues/16